shell/server/drivers/mail.js

```diff
@@ -223,15 +240,23 @@ hackSendEmail = (session, email) => {
       };
     }
 
-    if (email.from.address !== grainAddress && email.from.address !== userAddress.address) {
+    if (removeValidPlusSuffix(email.from.address) !== grainAddress && email.from.address !== userAddress.address) {
       throw new Error(
-        "FROM header in outgoing emails need to equal either " + grainAddress + " or " +
+        "FROM header in outgoing emails need to equal either " + grainAddress + " (with optional suffix) or " +
         userAddress.address + ". Yours was: " + email.from.address);
     }
 
     // Unpack fields
     const { from, to, cc, bcc, replyTo, subject, text, html } = email;
 
+    // We allow envelope address to have a suffix as well. In this way apps can use unique envelope
+    // addresses when sending out e-mails which allow better handling of e-mail bounces - app can
+    // match bounce with the sent e-mail.
+    let envelopeFrom = email.envelopeFrom || from.address;
```

Comment:
> Currently, envelopeFrom will never be present since it is not defined in email.capnp. There's no way for an app to fill this in.

Oh. I am not yet familiar enough with Sandstorm. But yes, then we have to extend the API a bit.

> if the app can't actually set an arbitrary envelopeFrom then it is confusing to let them specify one and then fail when they specify the wrong one.

I think this is pretty reasonable restriction, especially because we are restricting `From` as well. In the same sense it will fail if they set `From` to an arbitrary value.

> Maybe instead we should have a field bounceId which is strictly the ID to place after the +?

I think this is too specific. We can go with that, but then the same holds for `From`?

I think the current approach is the right balance between prescribing, restricting and allowing. Currently developer can choose:
* `From` as login e-mail address
* `From` as sandbox random e-mail address
* `From` as sandbox random e-mail address with custom suffix, if they want to expose the suffix to users (imagine mailing list service, where people should reply to the mailing list)
* any of the above for `From`, but customizing envelope from, to handle automatic replies in general

I think there is some advantage of envelope having same restrictions to  sandbox random e-mail address with optional custom suffix.

Anyway, it is your decision. I think we should try to overthink things here.

>  Another idea would be to extend the email interfaces such that Sandstorm can capture bounces and make some RPC callback to the app, which might be a nicer interface for apps, but requires a lot more work.

Yes, that is also an option.

> Are you actually anticipating the need to handle bounces in an app? If not then maybe we can punt this for now.

I wanted to use this primarily for sending notifications for posts to a forum out. And allowing recipients to reply to the e-mail to make a post. Similar to how GitHub allows replying to notifications about issues. Detecting bounces is useful to know issues with e-mails used in the forum.

But there are many open questions here if this is really the best approach. I have not yet checked new Sandstorm notifications API, so maybe that could be good enough. And there is also a limit on number of e-mails allowed to be send through Sandstorm. And also there is an issue of me needing to send e-mails in background, as a digest, when maybe I do not have user context/capability anymore. Also, e-mails used in the app should be all verified by Sandstorm already so bounces should be low.

So maybe at the end it will not work out, but this was the first step which I was thinking of improving in this aspect. I see it just as allowing more with a bit of code changes.